[SPARK-19911][STREAMING] Add builder interface for Kinesis DStreams

budde opened the pull request:
## What changes were proposed in this pull request?

- Add new KinesisDStream.scala containing KinesisDStream.Builder class
- Add KinesisDStreamBuilderSuite test suite
- Make KinesisInputDStream ctor args package private for testing
- Add JavaKinesisDStreamBuilderSuite test suite
- Add args to KinesisInputDStream and KinesisReceiver for optional
  service-specific auth (Kinesis, DynamoDB and CloudWatch)
## How was this patch tested?

Added ```KinesisDStreamBuilderSuite``` to verify builder class works as expected

SparkQA commented:
**[Test build #74342 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/74342/testReport)** for PR 17250 at commit [`8552caf`](https://github.com/apache/spark/commit/8552caf84dbd5b43a5a8446b10e4ccf9b295f924).

budde commented:
Open questions I'd like feedback on:

* Should the ```KinesisUtils.createStream()``` methods be marked as deprecated?
* Should the ```KinesisUtils.createStream()``` methods be refactored to use the builder?
* Should I add full docs for each method (e.g. including ```@param``` lists)?
* Does the file name and class name I've added seem reasonable?
* Is making the ctor args to ```KinesisInputDStream``` package private for testing reasonable?

I'd like to also extend this to allow configuring CloudWatch and DynamoDB-specific authorization which I imagine will be quite helpful to users. I'm trying to decide if I should do this as a separate PR or just roll it in here.

Pinging @brkyvz and @srown from #16744

SparkQA commented:
**[Test build #74342 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/74342/testReport)** for PR 17250 at commit [`8552caf`](https://github.com/apache/spark/commit/8552caf84dbd5b43a5a8446b10e4ccf9b295f924).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `  class Builder[T: ClassTag](`

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/74342/
Test FAILed.

budde commented:
Forgot to stop the ```StreamingContext``` added in ```KinesisDStreamBuilderSuite```. Updated the code to stop the context after all tests have run.

SparkQA commented:
**[Test build #74351 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/74351/testReport)** for PR 17250 at commit [`bcb7667`](https://github.com/apache/spark/commit/bcb76677b9fa58e5b96920d9faaa338183cca26d).

SparkQA commented:
**[Test build #74351 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/74351/testReport)** for PR 17250 at commit [`bcb7667`](https://github.com/apache/spark/commit/bcb76677b9fa58e5b96920d9faaa338183cca26d).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `  class Builder[T: ClassTag](`

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/74351/
Test PASSed.

Commit:
[SPARK-19911][STREAMING] Add builder interface for KinesisInputDStream

- Add KinesisInputDStream.Builder class
- Add KinesisInputDStreamBuilderSuite test suite
- Add JavaKinesisInputDStreamBuilderSuite test suite
- Rename SerializableCredentialsProvider -> SparkAWSCredentials
- Add SparkAWSCredentials.Builder
- Add SparkAWSCredentialsBuilderSuite test suite
- Make KinesisInputDStream ctor args package private for testing
- Add args to KinesisInputDStream and KinesisReceiver for optional
  service-specific auth (Kinesis, DynamoDB and CloudWatch)

SparkQA commented:
**[Test build #74379 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/74379/testReport)** for PR 17250 at commit [`a604dc5`](https://github.com/apache/spark/commit/a604dc5952a1c9939d433371abc969670a8ff6ab).

SparkQA commented:
**[Test build #74379 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/74379/testReport)** for PR 17250 at commit [`a604dc5`](https://github.com/apache/spark/commit/a604dc5952a1c9939d433371abc969670a8ff6ab).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `  class Builder[T: ClassTag](`

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/74379/
Test PASSed.

budde commented:
Found a bit more time to work on this. Changes made:

* Added implementations of ```KinesisDStream.builder()``` that take ```JavaStreamingContext``` instead of ```StreamingContext```
* Added ```JavaKinesisDStreamBuilderSuite``` to test that the builder interface is accessible in Java
* Added support for DynamoDB and CloudWatch-specific authentication parameters
* Made the documentation for ```KinesisDStream.Builder``` more thorough

SparkQA commented:
**[Test build #74418 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/74418/testReport)** for PR 17250 at commit [`8aeef08`](https://github.com/apache/spark/commit/8aeef08410999910e864d1de07f6eae21fd5409a).

SparkQA commented:
**[Test build #74418 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/74418/testReport)** for PR 17250 at commit [`8aeef08`](https://github.com/apache/spark/commit/8aeef08410999910e864d1de07f6eae21fd5409a).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `  class Builder[T: ClassTag](`

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/74418/
Test PASSed.

brkyvz commented on external/kinesis-asl/src/main/scala/org/apache/spark/streaming/kinesis/KinesisDStream.scala:
```diff
@@ -0,0 +1,559 @@
+     *
+     * @return Instance of [[KinesisInputDStream]] constructed with configured parameters
+     */
+    def build(): KinesisInputDStream[T] = new KinesisInputDStream(
```
I would get rid of the constructor params, and check for required values here, i.e.
 1. `ssc`
 2. `streamName`
 3. `appName`

budde replied:
I'll expand the builder to recognize required params and remove them as ctor args

brkyvz commented on external/kinesis-asl/src/main/scala/org/apache/spark/streaming/kinesis/KinesisDStream.scala:
```diff
@@ -0,0 +1,559 @@
+   * @return [[KinesisDStream.Builder]] instance
+   */
+  def builder[T: ClassTag](
+      jssc: JavaStreamingContext,
```
Have the builder take `setStreamingContext`, one which takes `StreamingContext` and the other `JavaStreamingContext`

budde replied:
Will be made required builder arguments

brkyvz commented on external/kinesis-asl/src/main/scala/org/apache/spark/streaming/kinesis/KinesisDStream.scala:
```diff
@@ -0,0 +1,559 @@
+  def builder[T: ClassTag](
+      jssc: JavaStreamingContext,
+      streamName: String,
+      messageHandler: Record => T): Builder[T] = builder(jssc.ssc, streamName, messageHandler)
```
let's also have a `setMessageHandler` function as well

budde replied:
Will do

brkyvz commented on external/kinesis-asl/src/main/scala/org/apache/spark/streaming/kinesis/KinesisDStream.scala:
```diff
@@ -0,0 +1,559 @@
+    = builder(jssc.ssc, streamName)
+
+  private[kinesis] val DEFAULT_KINESIS_ENDPOINT_URL: String
+    = "https://kinesis.us-east-1.amazonaws.com"
```
nit: move `=` to line above

budde replied:
Will do

brkyvz commented on external/kinesis-asl/src/main/scala/org/apache/spark/streaming/kinesis/KinesisInputDStream.scala:
```diff
@@ -31,15 +31,17 @@ import org.apache.spark.streaming.scheduler.ReceivedBlockInfo
+    private[kinesis] val initialPositionInStream: InitialPositionInStream,
+    private[kinesis] val checkpointAppName: String,
+    private[kinesis] val checkpointInterval: Duration,
+    private[kinesis] val _storageLevel: StorageLevel,
```
why the underscore?

budde replied:
Declaring ```val storageLevel``` collides with [```DStream.storageLevel```](https://github.com/apache/spark/blob/d5b1d5fc80153571c308130833d0c0774de62c92/streaming/src/main/scala/org/apache/spark/streaming/dstream/DStream.scala#L96)

brkyvz commented on external/kinesis-asl/src/main/scala/org/apache/spark/streaming/kinesis/KinesisDStream.scala:
```diff
@@ -0,0 +1,559 @@
+     * @param awsSecretKey AWS secret key for authenticating to STS
+     * @return Reference to this [[KinesisDStream.Builder]]
+     */
+    def kinesisSTSCredentials(
```
With this many methods for the credential provider, I feel we need a credential provider builder. I wouldn't want to re-enter everything again between `dynamoDb` and `cloudWatch`. If I want to keep them separate from the `kinesis` credentials

budde replied:
I'll look at introducing a credential builder

brkyvz reviewed (changes requested):
The direction is great. Major feedback is:
 1. Builder shouldn't have constructor params. It should check required values at the end.
 2. Please at `@deprecated` to `KinesisUtils.createStream` methods.
 3. We can have the `createStream` methods re-use the  builder pattern in a separate PR.
 4. We should think more about how to provide credentials. Ideally we only need 3 methods, not 3 x 6.
Maybe we should also have a `AWSCredentialsBuilder`. What do you think?


Ideally, the `Builder` class will be in `KinesisInputDStream`. `KinesisInputDStream` will have a `private constructor`, i.e. 
```scala
private[kinesis] KinesisInputDStream private (
  ...)
```

budde commented:
@brkyvz Thanks for taking a look!

Re: major feedback:

1. Sounds reasonable to me. I don't have strong feeling here.
1. Will do.
1. Sounds good.
1. I agree that method overloading isn't a great and could lead to a similar scenario as ```KinesisUtils.createStream()``` in the long run. I kept it here since I figured that the scope was rather limited and I wanted to avoid having to deal with too many builders but I think you raise a good point. I'll look into adding a credentials builder. Ideally we would only have to worry about Java/Python interoperability and could just use a single method optionals or default args to solve this but I guess we have to play the hands we're dealt :)

I can move the builder to the companion object of ```KinesisInputDStream``` (this is actually where I  had this placed originally). I like the idea of the private constructor as well although I don't think this will be possible until ```KinesisUtils.createStream()``` is refactored to use the builder pattern.

budde commented on external/kinesis-asl/src/main/scala/org/apache/spark/streaming/kinesis/KinesisDStream.scala:
```diff
@@ -0,0 +1,559 @@
+     *
+     * @return Reference to this [[KinesisDStream.Builder]]
+     */
+    def kinesisDefaultCredentials(): Builder[T] = {
```
Yeah, I mostly put this here for the sake of completeness. Should be able to remove this if we introduce a credential builder and replace it with a single method that just takes a ```SerializableCredentialsProvider``` instance to use for the service. A user could always just manually pass ```DefaultCredentialsProvider``` if they really want to for some reason.

budde commented on external/kinesis-asl/src/main/scala/org/apache/spark/streaming/kinesis/KinesisDStream.scala:
```diff
@@ -0,0 +1,559 @@
+    }
+
+    /**
+     * Sets the KCL application name to use when checkpointing state to DynamoDB. Defaults to the
```
We'll make this requited

budde commented on external/kinesis-asl/src/main/scala/org/apache/spark/streaming/kinesis/KinesisDStream.scala:
```diff
@@ -0,0 +1,559 @@
+
+    /**
+     * Sets the initial position data is read from in the Kinesis stream. Defaults to
+     * InitialPositionInStream.LATEST if no custom value is specified.
```
Will do

budde commented on external/kinesis-asl/src/main/scala/org/apache/spark/streaming/kinesis/KinesisDStream.scala:
```diff
@@ -0,0 +1,559 @@
+
+    /**
+     * Sets the AWS Kinesis endpoint URL. Defaults to "https://kinesis.us-east-1.amazonaws.com" if
+     * no custom value is specified
```
I'll take a look and see. To be honest supplying both of these has always felt pretty redundant to me. The AWS SDK has changed a bit in how it handles endpoints and regions as well, so it may also be worth revisiting how ```KinesisReceiver``` uses these params.

Long term, it may also be nice to allow for different endpoints to be specified for Kinesis, DynamoDB and CloudWatch (I think the KCL should support this...)

budde commented on external/kinesis-asl/src/main/scala/org/apache/spark/streaming/kinesis/KinesisDStream.scala:
```diff
@@ -0,0 +1,559 @@
+    private var endpointUrl = DEFAULT_KINESIS_ENDPOINT_URL
+    private var regionName = DEFAULT_KINESIS_REGION_NAME
+    private var initialPositionInStream = DEFAULT_INITIAL_POSITION_IN_STREAM
+    private var checkpointAppName = ssc.sparkContext.appName
```
Sounds reasonable. Feel free to push back on any other defaults as well-- figured these would just be a starting point.

budde commented on external/kinesis-asl/src/main/scala/org/apache/spark/streaming/kinesis/KinesisDStream.scala:
```diff
@@ -0,0 +1,559 @@
+   * @since 2.2.0
+   */
+  @InterfaceStability.Stable
+  class Builder[T: ClassTag](
```
This is probably the first builder class I've implemented so I'll defer to your judgment here :)

budde commented:
@brkyvz Actually, now that I think about it, do we need to make ```messageHandler``` a constructor arg since ```Builder``` is a generic class? There's probably a way we could get around this but I'd imagine it would be pretty complex...

brkyvz commented:
Good point @budde. I can think of two options:

 1. Leave it as a constructor param
 2. Make the `Builder` class non-generic and have the `build` function take the message handler:

```scala
class Builder {
  
  def build(): KinesisInputDStream[Array[Byte]]

  def buildWithMessageHandler[T](f: Record => T): KinesisInputDStream[T]
}
```

It's a matter of taking it as the first parameter or the final parameter. There are other ways to do it as well, but will throw runtime exceptions instead of at compile time.

cc @rxin for input on APIs

budde commented:
@brkyvz I think if we're eliminating the constructor arguments then the second approach you've proposed might make more sense. I can't think of anything cleaner.

brkyvz commented:
@budde Do you think you can update this PR? The 2.2 branch will be cut on Monday (2017-03-18).

budde commented:
@brkyvz A conference took up a lot of my time last week but I should have it updated later today

budde commented:
@brkyvz PR has been updated, apologies for the delay. I've added ```SerializableCredentialsProvider.Builder```, which I'm willing to hear suggestions for a better name on. I wanted to stay away from something like ```AWSCredentials.Builder``` so as to avoid confusion with similarly-named classes in the AWS Java SDK.

SparkQA commented:
**[Test build #74901 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/74901/testReport)** for PR 17250 at commit [`d6afaef`](https://github.com/apache/spark/commit/d6afaef4099d9b20d0d1257ec5942d1bb5b868af).

SparkQA commented:
**[Test build #74901 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/74901/testReport)** for PR 17250 at commit [`d6afaef`](https://github.com/apache/spark/commit/d6afaef4099d9b20d0d1257ec5942d1bb5b868af).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `  class Builder `
  * `  class Builder `

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/74901/
Test PASSed.

brkyvz commented on external/kinesis-asl/src/main/scala/org/apache/spark/streaming/kinesis/KinesisInputDStream.scala:
```diff
@@ -71,7 +75,256 @@ private[kinesis] class KinesisInputDStream[T: ClassTag](
+   */
+  def builder: Builder = new Builder
+
+  private[kinesis] def defaultMessageHandler(record: Record): Array[Byte] = {
```
would you mind just re-using the code in `KinesisUtils` instead of copying the code?

budde replied:
How about keeping it here and refactoring ```KinesisUtils``` to use it? I think this is what I was intending to do originally, just forgot to update the code.

brkyvz replied:
sounds good

SparkQA commented:
**[Test build #74909 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/74909/testReport)** for PR 17250 at commit [`3cc2df8`](https://github.com/apache/spark/commit/3cc2df82f5ae6b4e3a21c2549a8855f5bd8a3eb6).

budde commented:
@brkyvz Updated the PR to remove ```defaultMessageHandler()``` from ```KinesisUtils``` in favor of keeping this method in ```KinesisInputDStream```. My thought here was that this would be a better place for it since we've put ```KinesisUtils``` on the deprecation path.

SparkQA commented:
**[Test build #74909 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/74909/testReport)** for PR 17250 at commit [`3cc2df8`](https://github.com/apache/spark/commit/3cc2df82f5ae6b4e3a21c2549a8855f5bd8a3eb6).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `  class Builder `
  * `  class Builder `

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/74909/
Test PASSed.

brkyvz commented on external/kinesis-asl/src/main/scala/org/apache/spark/streaming/kinesis/SerializableCredentialsProvider.scala:
```diff
@@ -83,3 +84,146 @@ private[kinesis] final case class STSCredentialsProvider(
 }
+
+@InterfaceStability.Stable
+object SerializableCredentialsProvider {
```
hmm. This isn't a great name for a user facing API, the user shouldn't have to care about if the provider is serializable or not, that's an implementation detail.

I understand your concerns with the `AWSCredentials` name collisions. However, I think it's the best name there is.

brkyvz replied:
How about simply `CredentialProvider`

budde replied:
I agree we should definitely come up with a better name here. What about ```SparkAWSCredentials```? Obviously it's not as succinct as ```AWSCredentials``` but I think it's a clear name that avoids collisions.

I'm okay with ```CredentialsProvider``` otherwise.

brkyvz replied:
I guess `SparkAWSCredentials` also work

brkyvz commented on external/kinesis-asl/src/main/scala/org/apache/spark/streaming/kinesis/SerializableCredentialsProvider.scala:
```diff
@@ -83,3 +84,146 @@ private[kinesis] final case class STSCredentialsProvider(
+     * @param accessKeyId AWS access key ID
+     * @return Reference to this [[SerializableCredentialsProvider.Builder]]
+     */
+    def awsAccessKeyId(accessKeyId: String): Builder = {
```
anyone who provides an `accessKeyId` should also provide a `secretKey` therefore I would take both together.

`.withKeys(awsAccessKey: String, awsSecretKey: String)`

budde replied:
I'll rework this builder to take multiple arguments for the long-lived keypair and STS

brkyvz commented on external/kinesis-asl/src/main/scala/org/apache/spark/streaming/kinesis/SerializableCredentialsProvider.scala:
```diff
@@ -83,3 +84,146 @@ private[kinesis] final case class STSCredentialsProvider(
+     * @param roleArn ARN of IAM role to assume via STS
+     * @return Reference to this [[SerializableCredentialsProvider.Builder]]
+     */
+    def stsRoleArn(roleArn: String): Builder = {
```
same here: 
```scala
def withSts(roleArn: String, sessionName: String)
def withSts(roleArn: String, sessionName: String, externalId: String)
```

budde replied:
Will do

brkyvz commented on external/kinesis-asl/src/main/scala/org/apache/spark/streaming/kinesis/KinesisInputDStream.scala:
```diff
@@ -71,7 +75,256 @@ private[kinesis] class KinesisInputDStream[T: ClassTag](
+      throw new IllegalArgumentException(s"No value provided for required parameter $paramName")
+    }
+
+    private def buildSTSCredentials(
```
I guess you no longer need this

budde replied:
You're right, thanks for catching it

brkyvz commented on external/kinesis-asl/src/main/scala/org/apache/spark/streaming/kinesis/SerializableCredentialsProvider.scala:
```diff
@@ -83,3 +84,146 @@ private[kinesis] final case class STSCredentialsProvider(
   }
 }
+
+@InterfaceStability.Stable
```
I wouldn't call these `Stable` just yet :) Let's call it evolving for one release cycle

budde replied:
For sure. Honestly I just cribbed these annotations from ```SparkSession.Builder``` so I appreciate you letting me know what the proper convention is.

brkyvz commented on external/kinesis-asl/src/main/scala/org/apache/spark/streaming/kinesis/KinesisInputDStream.scala:
```diff
@@ -71,7 +75,256 @@ private[kinesis] class KinesisInputDStream[T: ClassTag](
   }
 }
+
+@InterfaceStability.Stable
```
ditto, let's call it evolving for now

budde replied:
Will do

brkyvz reviewed (commented):
Just a bit more to go.

I think we can rename `SerializableCredentialProvider` to `CredentialProvider`. Then it should have methods for keys and sts, not independent variables.

budde commented on external/kinesis-asl/src/main/scala/org/apache/spark/streaming/kinesis/KinesisInputDStream.scala:
```diff
@@ -71,7 +75,256 @@ private[kinesis] class KinesisInputDStream[T: ClassTag](
+    = "https://kinesis.us-east-1.amazonaws.com"
+  private[kinesis] val DEFAULT_KINESIS_REGION_NAME: String = "us-east-1"
+  private[kinesis] val DEFAULT_INITIAL_POSITION_IN_STREAM: InitialPositionInStream
+    = InitialPositionInStream.LATEST
```
Will fix. Sorry I keep doing this :-/

budde commented:
@brkyvz Updated per your feedback. Most significant change is renaming ```SerializableCredentialsProvider``` to ```SparkAWSCredentials``` (as well as renaming its subclasses) and refactoring its builder as you've suggested.

SparkQA commented:
**[Test build #75065 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/75065/testReport)** for PR 17250 at commit [`337b6ba`](https://github.com/apache/spark/commit/337b6ba575c969417566ee65664c93916e6923d4).

SparkQA commented:
**[Test build #75065 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/75065/testReport)** for PR 17250 at commit [`337b6ba`](https://github.com/apache/spark/commit/337b6ba575c969417566ee65664c93916e6923d4).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/75065/
Test PASSed.

brkyvz commented on external/kinesis-asl/src/main/scala/org/apache/spark/streaming/kinesis/KinesisInputDStream.scala:
```diff
@@ -22,24 +22,28 @@ import scala.reflect.ClassTag
-    storageLevel: StorageLevel,
-    messageHandler: Record => T,
-    kinesisCredsProvider: SerializableCredentialsProvider
+    private[kinesis] val streamName: String,
```
The `private`s are unnecessary since the class itself is private. You can either:
 1) remove `private[kinesis]`
 2) make it `private val` and then use `PrivateMethodTester` in the tests.

I'm fine either way.

budde replied:
Yeah, it's really just there so that I could access the values directly from the test. I'll look into using ```PrivateMethodTester```, thanks for the suggestion

brkyvz replied:
If you keep it `val` it should be fine.

budde replied:
Yeah, was just thinking that'd be a lot simpler. I'll go that route. Thanks!

brkyvz commented on external/kinesis-asl/src/main/scala/org/apache/spark/streaming/kinesis/KinesisInputDStream.scala:
```diff
@@ -71,7 +75,238 @@ private[kinesis] class KinesisInputDStream[T: ClassTag](
+     * Sets the [[SparkAWSCredentials]] to use for authenticating to the AWS Kinesis
+     * endpoint. Defaults to [[DefaultCredentialsProvider]] if no custom value is specified.
+     *
+     * @param credentials [[SerialziableCredentialsProvider]] to use for Kinesis authentication
```
`SparkAWSCredentials`

budde replied:
My bad, thanks

budde replied:
Figured out why I didn't catch this before-- apparently I can't spell "serializable". Ugh!

brkyvz commented on external/kinesis-asl/src/main/scala/org/apache/spark/streaming/kinesis/KinesisInputDStream.scala:
```diff
@@ -71,7 +75,238 @@ private[kinesis] class KinesisInputDStream[T: ClassTag](
+     * Sets the [[SparkAWSCredentials]] to use for authenticating to the AWS DynamoDB
+     * endpoint. Will use the same credentials used for AWS Kinesis if no custom value is set.
+     *
+     * @param credentials [[SerialziableCredentialsProvider]] to use for DynamoDB authentication
```
ditto

budde replied:
Will fix and do a "grep -r 'SerialziableCredentialsProvider' *" to make sure this isn't appearing anywhere else

brkyvz commented on external/kinesis-asl/src/main/scala/org/apache/spark/streaming/kinesis/KinesisInputDStream.scala:
```diff
@@ -71,7 +75,238 @@ private[kinesis] class KinesisInputDStream[T: ClassTag](
+     * Sets the [[SparkAWSCredentials]] to use for authenticating to the AWS CloudWatch
+     * endpoint. Will use the same credentials used for AWS Kinesis if no custom value is set.
+     *
+     * @param credentials [[SerialziableCredentialsProvider]] to use for CloudWatch authentication
```
ditto

brkyvz commented on external/kinesis-asl/src/test/java/org/apache/spark/streaming/kinesis/JavaKinesisInputDStreamBuilderSuite.java:
```diff
@@ -0,0 +1,63 @@
+    assert(kinesisDStream.checkpointAppName() == appName);
+    assert(kinesisDStream.checkpointInterval() == checkpointInterval);
+    assert(kinesisDStream._storageLevel() == storageLevel);
+    ssc.stop();
```
where does this come from? `LocalJavaStreamingContext`? If so, I wouldn't stop it. In fact, you can use `Mockito` to create a mock `JavaStreamingContext` if you like

budde replied:
I'll probably just go the mock route then and ignore the context all together. I was seeing a bunch of "Spark context is already running" error messages when I tried to run all of the streaming tests before I added this

budde replied:
Using a mock here and in the other test might not be very practical after all-- looks like the ```DStream``` constructor hooks into ```StreamingContext```. We would at least need to mock it's [```getState()``` method](https://github.com/apache/spark/blob/d5b1d5fc80153571c308130833d0c0774de62c92/streaming/src/main/scala/org/apache/spark/streaming/dstream/DStream.scala#L220) as well as [mocking a ```SparkContext``` along with its local properties](https://github.com/apache/spark/blob/d5b1d5fc80153571c308130833d0c0774de62c92/streaming/src/main/scala/org/apache/spark/streaming/dstream/DStream.scala#L220).

brkyvz replied:
that's fine then. As long as we don't break the environment for other tests, do proper clean up, it should be fine

budde replied:
Looks like this will need to be left as-is. In the current test implementation we check that  ```checkpointInterval``` isn't a required option and its default value is obtained via ```ssc.graph.batchDuration```, which we won't be able to mock because ```DStreamGraph``` is final.

brkyvz commented on external/kinesis-asl/src/test/scala/org/apache/spark/streaming/kinesis/KinesisInputDStreamBuilderSuite.scala:
```diff
@@ -0,0 +1,112 @@
+   with MockitoSugar {
+  import KinesisInputDStream._
+
+  private val ssc = new StreamingContext(conf, batchDuration)
```
since we're not really starting a stream, and testing API's, we should just mock it.

budde replied:
Will do

brkyvz commented on external/kinesis-asl/src/test/scala/org/apache/spark/streaming/kinesis/KinesisInputDStreamBuilderSuite.scala:
```diff
@@ -0,0 +1,112 @@
+    ssc.stop()
+  }
+
+  test("should raise an exception if any required field is missing") {
```
I would split this into 3 very small tests:
`should raise an exception if StreamingContext is missing`
`should raise an exception if stream name is missing`
`should raise an exception if checkpoint app name is missing`

budde replied:
Sounds good

brkyvz commented on external/kinesis-asl/src/test/scala/org/apache/spark/streaming/kinesis/SparkAWSCredentialsBuilderSuite.scala:
```diff
@@ -0,0 +1,112 @@
+        .build()
+    }
+    // No external ID, basic keypair for long-lived creds
+    assertResult(stsCreds.copy(
```
nit: could you make this a single line?

budde replied:
Will do

brkyvz commented on external/kinesis-asl/src/test/scala/org/apache/spark/streaming/kinesis/SparkAWSCredentialsBuilderSuite.scala:
```diff
@@ -0,0 +1,112 @@
+    }
+    // Default long-lived creds
+    assertResult(stsCreds.copy(
+      longLivedCredsProvider = DefaultCredentials)) {
```
nit: single line please.

budde replied:
I'll fix it. This happened since ```DefaultCredentialsProvider``` was shortened to ```DefaultCredentials``` so I'll try to check for other places where a multiline statement can be rolled up into a single line

brkyvz commented on external/kinesis-asl/src/test/scala/org/apache/spark/streaming/kinesis/SparkAWSCredentialsBuilderSuite.scala:
```diff
@@ -0,0 +1,112 @@
+  }
+
+  test("SparkAWSCredentials classes should be serializable") {
+    Utils.deserialize[BasicCredentials](
```
I know we didn't have it before, but could you also check that after deserialization, they're equivalent?
```scala
val creds = BasicCredentials("x", "y")
assert(Utils.deserialize[BasicCredentials](Utils.serialize(creds)) === creds)
```

budde replied:
Will do

brkyvz reviewed (commented):
Looking really good. I think we can merge it after this pass

budde commented:
@brkyvz Updated per your feedback, thanks for taking a thorough look. I also renamed the ```longLivedCredsProvider``` of ```STSCredentials``` to just ```longLivedCreds``` to match the updated naming conventions.

SparkQA commented:
**[Test build #75111 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/75111/testReport)** for PR 17250 at commit [`6f11978`](https://github.com/apache/spark/commit/6f11978326e59e4d383d7a64eef981b917364af6).

brkyvz commented:
Thanks a lot for the quick turnaround @budde ! Could you also contribute to the docs as well with the new builder API?
https://spark.apache.org/docs/latest/streaming-kinesis-integration.html

SparkQA commented:
**[Test build #75111 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/75111/testReport)** for PR 17250 at commit [`6f11978`](https://github.com/apache/spark/commit/6f11978326e59e4d383d7a64eef981b917364af6).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/75111/
Test PASSed.

budde commented:
@brkyvz Sure, want me to add it to this PR or open a new one?

brkyvz commented:
Thanks! new PR would be easier!

brkyvz commented on external/kinesis-asl/src/main/scala/org/apache/spark/streaming/kinesis/SparkAWSCredentials.scala:
```diff
@@ -0,0 +1,175 @@
+    private var stsCreds: Option[STSCredentials] = None
+
+    /**
+     * Use a basic AWS keypair for long-lived authorization.
```
Do you want to add the note here as well?
```
* @note The given AWS credentials will get saved in DStream checkpoints if checkpointing
* is enabled. Make sure that your checkpoint directory is secure.
```

budde replied:
Done

SparkQA commented:
**[Test build #75170 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/75170/testReport)** for PR 17250 at commit [`5315f1e`](https://github.com/apache/spark/commit/5315f1e9f200be7f97d7838c2d64bc6bfb208ecb).

brkyvz commented on external/kinesis-asl/src/main/scala/org/apache/spark/streaming/kinesis/SparkAWSCredentials.scala:
```diff
@@ -0,0 +1,179 @@
+     * Use a basic AWS keypair for long-lived authorization.
+     *
+     * @note The given AWS keypair will be saved in DStream checkpoints if checkpointing is
+     * enabled. Make sure that your checkpoint directory is secure if using the default
```
nit: `Make sure that your checkpoint directory is secure. Prefer using the [https://link.to.amazon.docs default credential provider chain]] if possible`

budde replied:
The link in this case will be quite long-- URL just by itself pushes it over the 100 char limit:

```[[http://docs.aws.amazon.com/sdk-for-java/v1/developer-guide/credentials.html#credentials-default default credential provider chain]]```

Do you know if there's a way to safely split this into multiple lines? Should I just turn style checks off for this comment?

brkyvz replied:
Feel free to add
```
// scalastyle: off

// scalastyle: on
```
around the doc

budde replied:
Done

SparkQA commented:
**[Test build #75172 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/75172/testReport)** for PR 17250 at commit [`03f91da`](https://github.com/apache/spark/commit/03f91dadb3878aa47f2a134e3e1b8d46aadd3b47).

brkyvz commented:
LGTM pending tests. Thanks a lot for this PR @budde

PS  It's okay to make new commits, you don't have to squash commit every time :)

SparkQA commented:
**[Test build #75170 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/75170/testReport)** for PR 17250 at commit [`5315f1e`](https://github.com/apache/spark/commit/5315f1e9f200be7f97d7838c2d64bc6bfb208ecb).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/75170/
Test PASSed.

SparkQA commented:
**[Test build #75172 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/75172/testReport)** for PR 17250 at commit [`03f91da`](https://github.com/apache/spark/commit/03f91dadb3878aa47f2a134e3e1b8d46aadd3b47).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/75172/
Test PASSed.

brkyvz commented:
Merging to master

budde commented:
@brkyvz Awesome, thanks for reviewing this!
